contracts/PrepaidCardMarket_v2.sol

```diff
@@ -0,0 +1,304 @@
+pragma solidity ^0.8.9;
+pragma abicoder v1;
+
+import "./core/Ownable.sol";
+
+import "@openzeppelin/contracts-upgradeable/utils/structs/EnumerableSetUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "./core/Versionable.sol";
+import "./IPrepaidCardMarket.sol";
+import "./PrepaidCardManager.sol";
+import "./ActionDispatcher.sol";
+import "./VersionManager.sol";
+import "./TokenManager.sol";
+import "./core/Safe.sol";
+import "@gnosis.pm/safe-contracts/contracts/proxies/GnosisSafeProxyFactory.sol";
+import "./libraries/SafeERC677.sol";
+
+import "hardhat/console.sol";
+
+contract PrepaidCardMarketV2 is Ownable, Versionable, Safe {
+  using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+  using SafeERC20Upgradeable for IERC677;
+  using SafeERC677 for IERC677;
+
+  struct SKU {
+    address issuerSafe;
+    address issuer;
+    address issuingToken;
+    uint256 faceValue;
+    string customizationDID;
+  }
+
+  address public prepaidCardManagerAddress;
+  address public tokenManager;
+  address public provisioner;
+  address public versionManager;
+  address public exchangeAddress;
+  mapping(address => mapping(address => uint256)) public balance; // issuer safe address -> token -> balance
```

Comment:
in order to support an issuer using multiple safes to fund their balance, maybe this should be //issuer EOA address -> token -> balance

Reply:
sorry, I take back, i think we should still track this at the safe level, otherwise you might be able to embezzle funds from a safe that has a threshold higher than 1 that you happen to be a part owner of